pkg/operator/encryption/state_test.go

```diff
@@ -0,0 +1,517 @@
+							},
+						},
+					},
+				}
+				ec := createEncryptionCfgWithWriteKey(keysRes)
+				return ec
+			}(),
+			output: map[schema.GroupResource]groupResourceKeys{
+				{Group: "", Resource: "secrets"}: {
+					writeKey: keyAndMode{
+						key: apiserverconfigv1.Key{Name: "34", Secret: "MTcxNTgyYTBmY2Q2YzVmZGI2NWNiZjVhM2U5MjQ5ZDc="}, mode: "aescbc",
+					},
+					readKeys: []keyAndMode{
+						{key: apiserverconfigv1.Key{Name: "33", Secret: "MTcxNTgyYTBmY2Q2YzVmZGI2NWNiZjVhM2U5MjQ5ZDc="}, mode: "aescbc"},
+					},
+				},
+				{Group: "", Resource: "configmaps"}: {
+					writeKey: keyAndMode{
+						key: apiserverconfigv1.Key{Name: "34", Secret: "MTcxNTgyYTBmY2Q2YzVmZGI2NWNiZjVhM2U5MjQ5ZDc="}, mode: "aescbc",
+					},
+					readKeys: []keyAndMode{
+						{key: apiserverconfigv1.Key{Name: "33", Secret: "MTcxNTgyYTBmY2Q2YzVmZGI2NWNiZjVhM2U5MjQ5ZDc="}, mode: "aescbc"},
+					},
+				},
+			},
+		},
+
+		// scenario 5
+		{
+			name: "single read key",
+			input: func() *apiserverconfigv1.EncryptionConfiguration {
+				ec := createEncryptionCfgNoWriteKey("34", "MTcxNTgyYTBmY2Q2YzVmZGI2NWNiZjVhM2U5MjQ5ZDc=", "secrets")
+				return ec
+			}(),
+			output: map[schema.GroupResource]groupResourceKeys{
+				{Group: "", Resource: "secrets"}: {
+					writeKey: keyAndMode{
+						key: apiserverconfigv1.Key{Name: "", Secret: ""}, mode: "identity",
+					},
+					readKeys: []keyAndMode{
+						{key: apiserverconfigv1.Key{Name: "34", Secret: "MTcxNTgyYTBmY2Q2YzVmZGI2NWNiZjVhM2U5MjQ5ZDc="}, mode: "aescbc"},
+					},
+				},
+			},
+		},
+
+		// scenario 6
+		{
+			name: "single read key multiple resources",
+			input: func() *apiserverconfigv1.EncryptionConfiguration {
+				ec := createEncryptionCfgNoWriteKey("34", "MTcxNTgyYTBmY2Q2YzVmZGI2NWNiZjVhM2U5MjQ5ZDc=", "secrets", "configmaps")
+				return ec
+			}(),
+			output: map[schema.GroupResource]groupResourceKeys{
+				{Group: "", Resource: "secrets"}: {
+					writeKey: keyAndMode{
+						key: apiserverconfigv1.Key{Name: "", Secret: ""}, mode: "identity",
+					},
+					readKeys: []keyAndMode{
+						{key: apiserverconfigv1.Key{Name: "34", Secret: "MTcxNTgyYTBmY2Q2YzVmZGI2NWNiZjVhM2U5MjQ5ZDc="}, mode: "aescbc"},
+					},
+				},
+				{Group: "", Resource: "configmaps"}: {
+					writeKey: keyAndMode{
+						key: apiserverconfigv1.Key{Name: "", Secret: ""}, mode: "identity",
+					},
+					readKeys: []keyAndMode{
+						{key: apiserverconfigv1.Key{Name: "34", Secret: "MTcxNTgyYTBmY2Q2YzVmZGI2NWNiZjVhM2U5MjQ5ZDc="}, mode: "aescbc"},
+					},
+				},
+			},
+		},
+
+		// scenario 7
+		{
+			name: "turn off encryption for single resource",
+			input: func() *apiserverconfigv1.EncryptionConfiguration {
+				keysRes := encryptionKeysResourceTuple{
+					resource: "secrets",
+					keys: []apiserverconfigv1.Key{
+						{
+							Name:   "34",
+							Secret: "MTcxNTgyYTBmY2Q2YzVmZGI2NWNiZjVhM2U5MjQ5ZDc=",
+						},
+
+						// secretsToProviders puts "fakeIdentityProvider" as last
```

Comment:
will this special key be ever removed?
what happens when a user frequently turns on and off the encryption?

Reply:
would be nice to add unit tests for ^

Reply:
I don't understand that well yet. Yes, at the very least we need heavy unit tests.